fix: use pv annotation to trigger filesystem resize when necessary

sunpa93 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
 /kind bug


**What this PR does / why we need it**:
fix: add annotation to pv after resize if filesystem resize is necessary. delete when pvc.status.capacity >= pv.spec.capacity

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/88683

**Special notes for your reviewer**:
This PR is a sister PR of https://github.com/kubernetes/kubernetes/pull/99326

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

gnufied commented:
/ok-to-test

gnufied commented:
/assign

gnufied commented on pkg/controller/controller.go:
```diff
@@ -533,11 +551,34 @@ func (ctrl *resizeController) patchClaim(oldPVC, newPVC *v1.PersistentVolumeClai
 
+	if fsResizeRequired {
+		// if the pv already has a resize annotation skip
+		if metav1.HasAnnotation(pv.ObjectMeta, util.AnnPreResizeCapacity) {
```
I think this is problematic - if I expand a PV from 10 to 20GB and while expansion is pending on the node, I expand it again to 40GB then second expansion won't be recorded in the pv spec?

sunpa93 replied:
You are definitely right on this. The function should not return before updating `pv.Spec.Capacity`

gnufied commented on pkg/controller/controller.go:
```diff
@@ -533,11 +551,34 @@ func (ctrl *resizeController) patchClaim(oldPVC, newPVC *v1.PersistentVolumeClai
+		if newPV.ObjectMeta.Annotations == nil {
+			newPV.ObjectMeta.Annotations = make(map[string]string)
+		}
+		newPV.ObjectMeta.Annotations[util.AnnPreResizeCapacity] = (&oldCapacity).String()
```
Why are we doing - `(&oldCapacity).String()` is that different from `oldCapacity.String()` ?

sunpa93 replied:
Done :)

gnufied reviewed (commented):
Looks good mostly. But I think we need unit tests for some these new behaviours.

sunpa93 commented:
> Looks good mostly. But I think we need unit tests for some these new behaviours.

I have modified the tests so that they check if annotations get properly populated and deleted

sunpa93 commented on pkg/controller/controller.go:
```diff
@@ -320,6 +321,16 @@ func (ctrl *resizeController) syncPVC(key string) error {
 		return fmt.Errorf("expected volume but got %+v", volumeObj)
 	}
 
+	if ctrl.isFsResizeComplete(pvc, pv) && metav1.HasAnnotation(pv.ObjectMeta, util.AnnPreResizeCapacity) {
```
I have renamed the function. I noticed that there are several places in the code that uses fsResize word. Will there be future issues to address this refactoring?

gnufied replied:
There is still conditions on the PVC which use fsResize wording, we can't rename that because that has become part of API now. But other than that we could and should rename everything else.

sunpa93 commented on pkg/controller/controller.go:
```diff
@@ -320,6 +321,16 @@ func (ctrl *resizeController) syncPVC(key string) error {
 
+	if ctrl.isFsResizeComplete(pvc, pv) && metav1.HasAnnotation(pv.ObjectMeta, util.AnnPreResizeCapacity) {
+		if err := ctrl.deletePreResizeCapAnnotation(pv); err != nil {
+			return fmt.Errorf("Failed removing annotation %s from pv %q: %v", util.AnnPreResizeCapacity, pv.Name, err)
```
Done :)

gnufied commented:
/lgtm

gnufied commented:
/assign @msau42

msau42 commented on pkg/controller/controller.go:
```diff
@@ -320,6 +321,16 @@ func (ctrl *resizeController) syncPVC(key string) error {
 	}
 
+	if ctrl.isNodeExpandComplete(pvc, pv) && metav1.HasAnnotation(pv.ObjectMeta, util.AnnPreResizeCapacity) {
+		if err := ctrl.deletePreResizeCapAnnotation(pv); err != nil {
```
how come resize controller is deleting the annotation? Shouldn't kubelet do it after nodeexpand is complete?

sunpa93 replied:
This is because kubelet does not have the permission to modify PV objects.

msau42 replied:
@gnufied is this an issue? Does this mean plugins that don't support controller resize and don't run the resizer will get this fix?

gnufied replied:
Currently I think external-resizer is required even for plugins that don't support controller expansion, because in that case external-resizer calls a NO-OP control-plane expansion and successfully expands the PV. There is no way to update PV spec from kubelet otherwise. So every plugin that supports expansion needs to run resizer.

gnufied commented on cmd/csi-resizer/main.go:
```diff
@@ -68,6 +68,8 @@ var (
 	handleVolumeInUseError = flag.Bool("handle-volume-inuse-error", true, "Flag to turn on/off capability to handle volume in use error in resizer controller. Defaults to true if not set.")
 
+	enableFSResizeAnnotation = flag.Bool("enable-fs-resize-annotation", true, "Flag to turn on/off capability to add an annotation to PV to denote its size prior to resize and to remove it upon completion of filesystem resize.")
+
```
I think we wanted this as a feature gate and not a flag.  The k8s feature gates look like - `--feature-gate=RestoreDeletedVolumeSize`

msau42 replied:
You can look at external-provisioner feature gate handling.

sunpa93 replied:
Done :)

Jiawei0227 commented:
Any updates on this? We are pending on csi-resizer release

gnufied commented on pkg/controller/controller.go:
```diff
@@ -21,6 +21,8 @@ import (
 	"fmt"
 	"time"
 
+	features "sigs.k8s.io/external-resizer/pkg/features" /* TO BE REPLACED WITH features "github.com/external-resizer/pkg/features once PR merged*/
```
Why are we importing via `sigs.k8s.io`? I do not understand reason for doing this. you should be able to import it normally via `github.com/kubernetes-csi/external-resizer`.

sunpa93 replied:
Done 👍

gnufied commented:
Can you squash 4 last commits in one?

sunpa93 commented:
> Can you squash 4 last commits in one?

Done :)

gnufied commented:
/lgtm

gnufied commented:
/approve
